drivers/rtc/rtc_rx8130ce.c

```diff
@@ -0,0 +1,757 @@
+/*
+ * Copyright (c) 2024 Måns Ansgariusson <mansgariusson@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(rx8130ce, CONFIG_RTC_LOG_LEVEL);
+
+#define DT_DRV_COMPAT epson_rx8130ce_rtc
+
+enum registers {
+	TIME		= 0x10,
+	ALARM		= 0x17,
+	/* control registers */
+	EXTENSION	= 0x1C,
+	FLAG		= 0x1D,
+	CTRL0		= 0x1E,
+	CTRL1		= 0x1F,
+	OFFSET		= 0x30,
+};
+
+#define RX8130CE_SECONDS_MASK	GENMASK(6, 0)
+#define RX8130CE_MINUTES_MASK	GENMASK(6, 0)
+#define RX8130CE_HOURS_MASK	GENMASK(5, 0)
+#define RX8130CE_DAYS_MASK	GENMASK(5, 0)
+#define RX8130CE_WEEKDAYS_MASK	GENMASK(6, 0)
+#define RX8130CE_MONTHS_MASK	GENMASK(4, 0)
+#define RX8130CE_YEARS_MASK	GENMASK(7, 0)
+
+#define RX8130CE_MONTHS_OFFSET (1)
+#define RX8130CE_YEARS_OFFSET (100)
+
+/* Alarm AE bit */
+#define ALARM_DISABLE	BIT(7)
+
+/* Extension reg(0x1C) bit field */
+#define EXT_TSEL0	BIT(0)
+#define EXT_TSEL1	BIT(1)
+#define EXT_TSEL2	BIT(2)
+#define EXT_WADA	BIT(3)
+#define EXT_TE		BIT(4)
+#define EXT_USEL	BIT(5)
+#define EXT_FSEL0	BIT(6)
+#define EXT_FSEL1	BIT(7)
+
+/* Flag reg(0x1D) bit field */
+#define FLAG_VBFF	BIT(0)
+#define FLAG_VLF	BIT(1)
+#define FLAG_RSF	BIT(2)
+#define FLAG_AF		BIT(3)
+#define FLAG_TF		BIT(4)
+#define FLAG_UF		BIT(5)
+#define FLAG_VBLF	BIT(7)
+
+/* Control0 reg(0x1E) bit field */
+#define CTRL0_TBKE	BIT(0)
+#define CTRL0_TBKON	BIT(1)
+#define CTRL0_TSTP	BIT(2)
+#define CTRL0_AIE	BIT(3)
+#define CTRL0_TIE	BIT(4)
+#define CTRL0_UIE	BIT(5)
+#define CTRL0_STOP	BIT(6)
+#define CTRL0_TEST	BIT(7)
+
+/* ctrl1 reg(0x1F) bit field */
+#define CTRL1_BFVSEL0	BIT(0)
+#define CTRL1_BFVSEL1	BIT(1)
+#define CTRL1_RSVSEL	BIT(2)
+#define CTRL1_INIEN	BIT(4)
+#define CTRL1_CHGEN	BIT(5)
+#define CTRL1_SMPTSEL0	BIT(6)
+#define CTRL1_SMPTSEL1	BIT(7)
+
+
+/* Digital Offest reg(0x30) bit field */
+#define DIGITAL_OFFSET_NEG	BIT(6)
+#define DIGITAL_OFFSET_DTE	BIT(7)
+
+/* Digital Offset register values */
+#define DIGITAL_OFFSET_MAX	192260
+#define DIGITAL_OFFSET_MIN	-195310
+#define DIGITAL_OFFSET_STEP_PPB	3050
+
+/**
+ * @brief rx8130ce control registers
+ * 0x1C extension register
+ * 0x1D Flag register
+ * 0x1E control0
+ * 0x1F ctrl1
+ */
+struct __packed rx8130ce_registers {
+	uint8_t extension;
+	uint8_t flag;
+	uint8_t ctrl0;
+	uint8_t ctrl1;
+};
+
+struct __packed rx8130ce_time {
+	uint8_t second;
+	uint8_t minute;
+	uint8_t hour;
+	uint8_t weekday;
+	uint8_t day;
+	uint8_t month;
+	uint8_t year;
+};
+
+struct __packed rx8130ce_alarm {
+	uint8_t minute;
+	uint8_t hour;
+	union {
+		uint8_t wday;
+		uint8_t day;
+	};
+};
+
+struct rx8130ce_config {
+	const struct i2c_dt_spec i2c;
+	struct gpio_dt_spec irq;
+	uint8_t clockout_frequency;
+	uint8_t battery_switchover;
+};
+
+struct rx8130ce_data {
+	struct k_mutex lock;
```

Comment:
Please avoid mutexes for locking, they are very expensive compared to a binary semaphore, and only required if thread ownership needs to be tracked:
```
k_sem_take();
k_sem_give();
```

Reply:
In `rx8130ce_irq_work_handler` the driver own the lock when the driver call the callback function(s) to the user. 

If I switch to a semaphore the callbacks would be unable to read time or set new alarms so I opt to keep it as is.

Reply:
Driver callbacks are usually not meant to be used to do work, most of them are called from interrupt contexts, including integrated RTCs. If you want the driver APIs to be callable from the callback, including setting or clearing the callback itself, its simpler to copy the callback and user data in `rx8130ce_irq_work_handler,` release the lock, then call the callback.

Reply:
Done :+1: